crates/amaru-ledger/src/rules/transaction/collateral.rs

```diff
@@ -0,0 +1,363 @@
+    #[error("too many collateral inputs: provided: {provided} allowed: {allowed}")]
+    TooManyInputs { provided: usize, allowed: usize },
+    #[error("a collateral input is locked at a script address: {0}")]
+    LockedAtScriptAddress(TransactionInputAdapter),
+    #[error("total collateral value is insufficient: provided: {provided} required: {required}")]
+    InsufficientBalance { provided: u64, required: u64 },
+    #[error(
+        "total collateral field (expected) does not equal actual collateral (provided): provided: {provided} expected: {expected} "
+    )]
+    IncorrectTotalCollateral { provided: u64, expected: u64 },
+    #[error("No collateral was provided, but collateral is required")]
+    NoCollateral,
+    #[error("collateral has non-zero delta: {0:?}")]
+    ValueNotConserved(CollateralBalance),
+    // TODO: This error shouldn't exist, it's a placeholder for better error handling in less straight forward cases
+    #[error("uncategorized error: {0}")]
+    UncategorizedError(String),
+}
+
+/*
+ Collateral validation occurs during fee validation in the Haskell node. See the comments below for notes on collateral validation:
+ https://github.com/IntersectMBO/cardano-ledger/blob/master/eras/babbage/impl/src/Cardano/Ledger/Babbage/Rules/Utxo.hs#L180-L195
+*/
+pub fn execute<C>(
+    context: &mut C,
+    collaterals: Option<&[TransactionInput]>,
+    collateral_return: Option<&MintedTransactionOutput<'_>>,
+    tx_collateral: Option<u64>,
+    fee: u64,
+    protocol_parameters: &ProtocolParameters,
+) -> Result<(), InvalidCollateral>
+where
+    C: UtxoSlice,
+{
+    let collaterals = collaterals
+        .filter(|c| !c.is_empty())
+        .ok_or(InvalidCollateral::NoCollateral)?;
+
+    let mut balance = CollateralBalance {
+        coin: 0,
+        multiasset: BTreeMap::new(),
+    };
+
+    let allowed = protocol_parameters.max_collateral_inputs as usize;
+    let provided = collaterals.len();
+    if provided > allowed {
+        return Err(InvalidCollateral::TooManyInputs { provided, allowed });
+    }
+
+    for collateral in collaterals.iter() {
+        let output = context
+            .lookup(collateral)
+            .ok_or_else(|| InvalidCollateral::UnknownInput(collateral.into()))?;
+
+        if output.address.has_script() {
+            return Err(InvalidCollateral::LockedAtScriptAddress(collateral.into()));
+        }
+
+        add_value_to_balance(&mut balance, output);
+    }
+
+    let collateral_return_balance = match collateral_return {
+        Some(output) => match output {
+            amaru_kernel::PseudoTransactionOutput::Legacy(output) => {
+                CollateralBalance::from(&output.amount)
+            }
+
+            amaru_kernel::PseudoTransactionOutput::PostAlonzo(output) => {
+                CollateralBalance::from(&output.value)
+            }
+        },
+        None => CollateralBalance {
+            coin: 0,
+            multiasset: BTreeMap::new(),
+        },
+    };
```

Comment:
That code is too low-level for rules; please refactor so that the pattern-match lives inside kernel's helper functions or CollateralBalance's helper function.